
title: "Import and normalise Maryland CC"
author: "Asbj√∏rn"
date: "2022/2/24"
output: html_notebook
editor_options: 
  chunk_output_type: inline
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
library(tidyverse)
#library(annotables)
library(edgeR)
library(RColorBrewer)
library(ggrepel)
library(limma)
library(magrittr)
library(biomaRt)
theme_set(theme_bw()) #change default ggplot background
theme_update(text= element_text(size = 7, family = "sans"))
#fixing default function names
select <- dplyr::select
rename <- dplyr::rename
filter <- dplyr::filter
relocate <- dplyr::relocate
colnames <- base::colnames
rownames <- base::rownames
```


This code is to import the Maryland data and produce normalised and filtered data.



### Add sample info

```{r add info}
info_OFC <- 
  readxl::read_xlsx("/Users/asbjorh/PhD/ANK3_HTS-amplicons/Maryland/data/Maryland_Samples_NBB1582DonorInfo_cleaned.xlsx", sheet = 1, col_names = TRUE) %>% 
  dplyr::select(c("RNA ID", "AGEYEARS", "AGEDAYS", "RIN AMG", "RNAseq", "PMINTERVAL", "CDEATHOFF")) %>% filter(RNAseq!=0) %>% select(-RNAseq) %>% distinct()

#simplify the names of samples to match data
info_OFC$`RNA ID` <-  str_remove(info_OFC$`RNA ID`, "MD_")
info_OFC$`RNA ID` <-  str_remove(info_OFC$`RNA ID`, "_RNA")
info_OFC$`RNA ID` <-  str_replace(info_OFC$`RNA ID`, "_", "-")
info_OFC <- dplyr::rename(info_OFC, sampleID="RNA ID")

info_OFC <- info_OFC %>% filter(!(sampleID=="5162-BA11" & `RIN AMG`== 9.6)) #this sample has 2 entries, unknown which was used for RNAseq, remove best RIN, keep RIN = 9.1

info_OFC <- dplyr::rename(info_OFC, RIN_AMG="RIN AMG")

info_OFC$sampleID <- str_remove(info_OFC$sampleID,  "-BA11")

info_CC <-
readxl::read_xlsx("~/proj_Maryland_WGCNA/data/RNAseq_CC samples_december_2021/sample_table_96_RNAseq_cleaned.xlsx", sheet = 1, col_names = TRUE)


info_CC <- 
  info_CC %>% 
  select(sampleID, RIN_AMG)

info_CC$sampleID <- as.character(info_CC$sampleID)

info_CC <- 
  info_CC %>% left_join(select(info_OFC, -RIN_AMG), by= "sampleID")


#change age (years+days) into decimal years. Remove cause of death.

info_CC <- 
  info_CC %>% 
  mutate(sampleID = paste0(sampleID, "-CC")) %>% 
  dplyr::mutate(AGE = AGEYEARS + (AGEDAYS/365)) %>%
    select(-AGEYEARS, -AGEDAYS)

info_OFC <- 
  info_OFC %>% 
  mutate(sampleID = paste0(sampleID, "-OFC")) %>% 
  select(-CDEATHOFF) %>% 
  dplyr::mutate(AGE = AGEYEARS + (AGEDAYS/365)) %>%
    select(-AGEYEARS, -AGEDAYS)

#mke a category for age. IA suggested +-12 Years, based on Kang et al (2011)'s definition of adolescence 

info_CC <-
  info_CC %>% 
  mutate(below12 = case_when(
    AGE < 12  ~ 1,
    TRUE ~ 0)
  )

info_CC$below12 <- as.factor(info_CC$below12)

#do the same for OFC
info_OFC <-
  info_OFC %>% 
  mutate(below12 = case_when(
    AGE < 12  ~ 1,
    TRUE ~ 0)
  )

info_OFC$below12 <- as.factor(info_OFC$below12)
```



```{r import counts CC}
#f = list.files("/Users/asbjorh/PhD/RNAseq_temporal/data/CC/Hughes-RNA6-2021-12-08/20_mapDataRSEM/",  pattern= ".genes.results$", full.names  = TRUE) #results from RSEM

f = list.files("/Users/asbjorh/PhD/RNAseq_temporal/data/CC/Hughes-RNA6-2021-12-08/40_countData/",  pattern= ".counts$", full.names  = TRUE) #results from 


#if RSEM results:
#dat = lapply(f, 
#             function(i){          #get the data from each sample
#  x = read_tsv(i, col_names = TRUE, skip = 0)
#  x = x[, c(1,4,5)] 
#  # add a column to say which file they're from
#  x$file = i
#  # Return your data
#  x
#}
#)

#if featurecounts results:
dat = lapply(f, 
             function(i){          #get the data from each sample
  x = read_tsv(i, col_names = FALSE, skip = 2)
  x = x[, c(1,6,7)] 
  # add a column to say which file they're from
  x$file = i
  # Return your data
  x
}
)


d<-
  do.call("rbind.data.frame", dat) #make a dataframe out of the imported datasets (combine by )


d$file <- #remove unnescessary prefix text from sample names
  str_remove(d$file, "/Users/asbjorh/PhD/RNAseq_temporal/data/CC/Hughes-RNA6-2021-12-08/40_countData//") 

d$file  <- #replace unnescessary suffix text from sample names
  str_replace(d$file , ".counts", "-CC") 


d <- 
  pivot_wider(dplyr::select(d, -X6), names_from= file, values_from = X7 ) 

 count <- d %>% column_to_rownames(var="X1")



#if using RSEM data
#d$file <- #remove unnescessary prefix text from sample names
 # str_remove(d$file, "/Users/asbjorh/PhD/RNAseq_temporal/data/CC/Hughes-RNA6-2021-12-08/20_mapDataRSEM//") 

#d$file  <- #replace unnescessary suffix text from sample names
 # str_replace(d$file , "_rsemExpr.genes.results", "-CC") 

#count <- d %>% select(gene_id, file, expected_count) %>% pivot_wider(names_from = file, values_from = expected_count, id_cols = gene_id) 
#count <- count %>% column_to_rownames(var="gene_id")


```

```{r import counts OFC}
f = list.files("/Users/asbjorh/PhD/RNAseq_temporal/data/OFC/",  pattern= ".counts$", full.names  = TRUE) #results from 


#if featurecounts results:
dat = lapply(f, 
             function(i){          #get the data from each sample
  x = read_tsv(i, col_names = FALSE, skip = 2)
  x = x[, c(1,6,7)] 
  # add a column to say which file they're from
  x$file = i
  # Return your data
  x
}
)


d<-
  do.call("rbind.data.frame", dat) #make a dataframe out of the imported datasets (combine by )


d$file <- #remove unnescessary prefix text from sample names
  str_remove(d$file, "/Users/asbjorh/PhD/RNAseq_temporal/data/OFC//") 

d$file  <- #replace unnescessary suffix text from sample names
  str_replace(d$file , ".counts", "-OFC") 


d <- 
  pivot_wider(dplyr::select(d, -X6), names_from= file, values_from = X7 ) 

 count <- d %>% column_to_rownames(var="X1")

 colnames(count) <- str_remove(colnames(count),  "-BA11")

```




```{r Process data}
clinMeta <- info_CC
#clinMeta <- info_OFC

# exclude outlier CC_1259 sample
exclude <- "1259-CC"
count <- count

 count <- count[, !colnames(count) %in% exclude] 

SampleID <- colnames(count)[colnames(count) %in% clinMeta$sampleID] #make any other adjustments to which samples to include/ exclude here

data <- count[,colnames(count) %in% (SampleID)] #count is already in the form of a dataframe with rownames
data <- round(data) # round data since RSEM output may be non-integer

colData <- clinMeta


colData <- colData %>%  column_to_rownames(var="sampleID") #to make a dataframe with rownames
colData <- colData[colnames(data),] #order the sample names the same as the expression data
all( rownames(colData) == colnames(data))

#colData[,c(2,4,5,6)] <- lapply(colData[,c(2,4,5,6)],as.factor) #adjust any columns with factor values
#colData[,c(3,7,8)] <- lapply(colData[,c(3,7,8)],as.numeric) #similarly, make sure numerical values are properly labeled.


library(biomaRt)
ensembl <- useMart("ensembl", dataset="hsapiens_gene_ensembl", host="https://useast.ensembl.org")
genemap <- getBM(attributes=c("entrezgene_id","ensembl_gene_id", "external_gene_name", "gene_biotype", "chromosome_name", "start_position", "end_position"), filters="ensembl_gene_id", 
	values=rownames(data), mart=ensembl)
genemap <- genemap[ duplicated(genemap$ensembl_gene_id) == F,]
rownames(genemap) <- genemap$ensembl_gene_id #rownames in case dataframe is needed


## Remove lowly expressed genes

library(edgeR)
keep <- filterByExpr(data#, design=colData$below12 #we don't want to have design here, do we? Because if genes are not at all expressed in either category it will be filtered out?
)
data.filt <- data[keep,]


## Keep only protein-coding and lncRNA genes
genemap2 <- genemap[rownames(data.filt),]
genemap2 <- subset(genemap2, gene_biotype %in% c("protein_coding","lncRNA"))
data.filt <- data.filt[rownames(genemap2),]
```


```{r DE analyses childhood vs adolescence}
analysis <- "CC_DE_below12"
#analysis <- "OFC_DE_below12"

res_dir <- paste0("~/proj_Maryland_WGCNA/analysis/", analysis)


y <- DGEList(counts = data.filt, genes = genemap2, samples=SampleID)
y <- calcNormFactors(y)

design <- model.matrix(~ below12  + RIN_AMG + PMINTERVAL, data = colData)

v <- voomWithQualityWeights(y, design, plot = TRUE)
fit <- lmFit(v, design)
fit <- eBayes(fit)
summary(decideTests(fit, p.value=0.1))

res <- topTable(fit, coef="below121", n=Inf, p.value=1)
write.table(res, paste0(res_dir, "/de_limma_OFC_below12.txt"), sep="\t", row.names=F)
```


```{r plot DE genes}
#in order to visualise expression, we need the samples as rows and gene expression + covariates as columns
## Plot significant genes
library(ggsci)
library(ggpubr)

col <- get_palette("aaas", 10)
colData.filt <- subset(colData, rownames(colData) %in% SampleID )


df <- cpm(y, log=TRUE, prior.count=3) #convert counts to normalised cpm
df <- df[rownames(head(res,10)), ] #only keep the top 10 significant genes

df <- df %>% t() %>% data.frame()     #have genes as columns
colnames(df) <- head(res,10)$external_gene_name #use gene symbols in stead of ensembl codes
df <- merge(df,colData.filt,by=0) 

df <- df %>% rename("SampleID" = Row.names )


#IA's plotting
# glist <- lapply(2:11, function (x) {
#   ggplot() + 
#     geom_boxplot(data=df,aes(x = below12, y = df[,x]), color ="black",size=0.7, outlier.shape=NA, fill="white") +
#     geom_point(data=df,aes(x = below12, y = df[,x], color = below12, fill=below12), position=position_jitterdodge(), pch=21,size=3.5, color="black", alpha=0.85) + 
#     labs(title=colnames(df)[x],x="",y="") +
#     scale_colour_manual(values=col[c(2,9,8)]) +
#     scale_fill_manual(values=col[c(2,9,8)]) +
#     theme_classic() +
#     theme(axis.text.x=element_text(size=20,color="black", angle=0, hjust=0.5),
#           axis.text.y=element_text(size=20,color="black"),
#           axis.title=element_text(size=16),
#           plot.title=element_text(size=20,hjust=0.5),
#           legend.text=element_text(size=20),
#           legend.title=element_blank(),
#           plot.margin=unit(c(0,0,0,0),"cm"),
#           panel.background=element_rect(linetype="solid", color="black", size=1)) 
# })
# 
# 
# annotate_figure(ggarrange(plotlist=glist,ncol=5,nrow=2,common.legend=T,legend="none"),
#                 left=text_grob("Normalized expression",size=20,rot=90))
# 

#prepare for flexible plotting
df_long <- 
  df %>%  pivot_longer(cols = 2:11, names_to = "gene_id", values_to = "log_cpm") 

df_long$below12 <- str_replace(df_long$below12, "1", "<12")
df_long$below12 <- str_replace(df_long$below12, "0", ">12")

df_long$below12 <- factor(df_long$below12, levels =  c("<12", ">12"))

df_long <- df_long %>%  rename("Age_group" = below12)

plot_expr <- df_long %>% ggplot(aes(x=Age_group, y= log_cpm)) + 
    geom_boxplot( color ="black",size=0.7, outlier.shape=NA, fill="white") +
    geom_point(aes( colour= Age_group, fill = Age_group), position= "jitter") +
  facet_wrap(~gene_id, scales = "free_y") +
      scale_colour_manual(values=col[c(1,6)]) +
    scale_fill_manual(values=col[c(1,6)]) +
  labs(title= paste0("Normalised expression, 10 significant DE genes, ", analysis)) +
  theme_bw()

ggsave(plot_expr, file = paste0(res_dir, "/plot_expr_top10sign.pdf" ))  
  

```


```{r variance in age groups}
library(rstatix)

DE_genes_0.05 <- 
  res %>% rownames_to_column(var = "gene_id") %>% filter(adj.P.Val < 0.05) %>% select(external_gene_name)

#look at 10 most significant genes

df_long %>% 
  select(SampleID, Age_group, gene_id, log_cpm) %>% 
  group_by(Age_group, gene_id) %>% 
  summarise(sd = sd(log_cpm)) %>% 
  ungroup() %>% 
  ggplot(aes(x= Age_group, y= sd)) + geom_boxplot() + geom_jitter() #%>% 
  #t_test(sd  ~ Age_group)
 
#interesting trend. Should we look for genes with high difference in variance? 
cpm_all <- cpm(y, log=TRUE, prior.count=3) #convert counts to normalised cpm

cpm_all <- cpm_all %>% t() %>% data.frame()     #have genes as columns
cpm_all_long <- 
  cpm_all %>%  rownames_to_column(var = "SampleID") %>% 
  pivot_longer(cols = -1, names_to = "ensembl_gene_id", values_to = "log_cpm")

#change gene name to symbol
cpm_all_long <- 
  cpm_all_long %>% 
  left_join(select(res, ensembl_gene_id, external_gene_name) , by=c("ensembl_gene_id"))  #add in symbol name

cpm_all_long <- 
  cpm_all_long %>% rename("gene_id" = external_gene_name)

cpm_all_long <- 
  cpm_all_long %>% 
  left_join(rownames_to_column(colData.filt, var= "SampleID"), by="SampleID")

cpm_all_long$below12 <- str_replace(cpm_all_long$below12, "1", "<12")
cpm_all_long$below12 <- str_replace(cpm_all_long$below12, "0", ">12")
cpm_all_long$below12 <- factor(cpm_all_long$below12, levels =  c("<12", ">12"))
cpm_all_long <- cpm_all_long %>%  rename("Age_group" = below12)

genes_sd_diff <- 
  cpm_all_long %>% 
  select(SampleID, Age_group, gene_id, log_cpm) %>% 
  group_by(Age_group, gene_id) %>% 
  summarise(sd = sd(log_cpm)) %>% 
  ungroup() %>% 
  pivot_wider(id_cols = gene_id, names_from = Age_group, values_from = sd) %>% 
  mutate(sd_diff = `<12` - `>12`) %>% 
  arrange(desc(sd_diff)) %>% 
  head(n= 20) %>% 
  select(gene_id)

#plot_expr <- 
  cpm_all_long %>% 
    filter(gene_id %in% genes_sd_diff$gene_id) %>% 
    ggplot(aes(x=Age_group, y= log_cpm)) + 
    geom_boxplot(  outlier.shape=NA, fill="white") +
    geom_point(aes( colour = AGE), position= "jitter") +
  facet_wrap(~gene_id, scales = "free_y") +
 #     scale_colour_manual(values=col[c(1,6)]) +
 #   scale_fill_manual(values=col[c(1,6)]) +
  #labs(title= paste0("Normalised expression, 20 genes with high differense in SD, ", analysis)) +
  theme_bw()

  #it is clear that the largest difference in sd is due to 1 outlier sample in many cases, and it is not the AGE<2 samples
  


```

```{r pairs of switches}
#Kang et al
#IGF2 in fig 2c
#DCX in fig 5d


#Bar-Shira et al's selection:
#GRIN2A/GRIN2B #mostly embryonic
#GABRA3/ GABRA6
#GLRA2/GLRA3
#HTR1A/ HTR5A

  cpm_all_long %>% 
 #   filter(gene_id %in% c("GRIN2A", "GRIN2B")) %>% 
#      filter(gene_id %in% c("GLRA2", "GLRA3")) %>% 
#        filter(gene_id %in% c("HTR1A", "HTR5A")) %>% 
  filter(gene_id %in% c("GABRA3", "GABRA6")) %>% 
    ggplot(aes(x=AGE, y= log_cpm, colour=gene_id)) +
  geom_point()+
  geom_smooth(se = FALSE) +
        scale_colour_manual(values=col[c(1,6)])


```


```{r find our own switch pairs}
#Bar-Shira et al 2015 call these potential switch pairs "condition-dependent variants" or CDVs

# Specifically, we used the KEGG
# pathways repository [2] to focus on proteins that participate in synaptic pathways, including
# signal transduction pathways of all neurotransmitters (a total of 17 KEGG pathways). We then
# used KEGG to define subgroups of functionally-related proteins which are likely to contain
# CDVs (see Methods). Then, we calculated the similarity of the two amino-acid sequences of
# each candidate pair, and kept pairs with sequence similarity above 30% (see Methods). Low
# sequence similarity is usually associated with different functionality and dissimilar protein
# sequences can be excluded from consideration [40]

library(KEGGREST)
library(org.Hs.eg.db)

#the following KEGG codes are linked to neuronal functions, neurodegenerative disorders and substance dependence. They are taken from Table2 in Bar-Shira et al 2015. 
neuronal_KEGG <- c(
  "path:hsa04020",
  "path:hsa04724",
  "path:hsa04725",
  "path:hsa04726",
  "path:hsa04727",
  "path:hsa04728",
  "path:hsa04730",
  "path:hsa05010",
  "path:hsa05012",
  "path:hsa05014",
  "path:hsa05016",
  "path:hsa05030",
  "path:hsa05031",
  "path:hsa05032",
  "path:hsa05033",
  "path:hsa05034",
    "path:hsa04080"
)

#get the entrez ID from genes that are in these KEGG pathways
hsa_path_eg  <- 
  keggLink(target = "hsa", source = neuronal_KEGG) %>%
  tibble(gene = ., KEGG_pathway = sub("hsa:", "", names(.))) %>% 
   separate_wider_delim(gene, delim = ":", names = c(NA, "entrezgene_id")) %>% 
  mutate(KEGG_pathway= str_replace_all(hsa_path_eg$KEGG_pathway, "path:hsa", "")) 

#entrez stored as numbers by default. Need character. 
genemap2$entrezgene_id <- as.character(genemap2$entrezgene_id) 

#get more gene info. genemap2 is already reduced to protein-coding and lncRNA
hsa_path_eg <- 
  hsa_path_eg %>% 
  left_join(select(tibble(genemap2), entrezgene_id, ensembl_gene_id, external_gene_name), by= "entrezgene_id") %>% 
  tidyr::drop_na()

#of neuro-KEGG genes, find paralogs
paralogs_neuro <- getBM(
  attributes = c( "ensembl_gene_id", "external_gene_name", "hsapiens_paralog_ensembl_gene", "hsapiens_paralog_associated_gene_name" , "hsapiens_paralog_perc_id"), 
  mart= ensembl, 
  filters = "external_gene_name", 
  values = hsa_path_eg$external_gene_name)

paralogs_neuro <-
  paralogs_neuro %>% tidyr::drop_na()

#Bar-Shira et al used 30% sequence similarity as cut-off for possible CDVs
paralogs_neuro_sim <-
  paralogs_neuro %>%
  filter(hsapiens_paralog_perc_id>= 30) 
#this is gene similarity, while the paper used "protein pairs with amino acid sequence similarity lower than 30%";
# Protein sequences were aligned using the Needleman-Wunsch global algorithm. Following the
# definition used for global alignment in BLAST 2 Sequences [78] we used blosum62 as the scoring
# matrix, and set the gap costs to 11 for gap existence and 1 for each extension. The sequence
# similarity was defined as the fraction of identical residues out of the number of residues in the
# longer sequence.

#Quantify dissimilarity between expression profiles
#the Spearman correlation between the expression profiles of each pair of genes. Analysis was limited to differentially expressed genes, by excluding genes whose expression range (maximum‚Äìminimum) was lower than 1.5 in log2 scale

#find genes that does not vary much over time, and exclude them from analysis
genes_range1.5<- 
  cpm_all_long %>% 
  select(SampleID, log_cpm, gene_id) %>% 
  group_by(gene_id) %>% 
  mutate(max= max(log_cpm), min= min(log_cpm)) %>% 
  mutate(range= max-min) %>% ungroup() %>% 
  select(gene_id, range) %>% distinct() %>%  
  filter(range>1.5) %>% 
  select(gene_id)

paralogs_neuro_sim <- 
  filter(paralogs_neuro_sim, external_gene_name %in% genes_range1.5$gene_id)

#Simple method: find pearson correlation of pairs, and look for negatively correlated expression

cor <- cpm_all %>% 
  dplyr::select(colnames(cpm_all)[colnames(cpm_all) %in% paralogs_neuro_sim$ensembl_gene_id] | colnames(cpm_all)[colnames(cpm_all) %in% paralogs_neuro_sim$hsapiens_paralog_ensembl_gene] ) %>% 
  cor()                    # get all correlations of neuronal genes with paralogs or that are paralogs of these 
  

cor[lower.tri(cor)] <- NA #removes the reverse correlation

cor_paralogs_neuro_sim <- 
  cor %>% 
 data.frame() %>%           # save result as a dataframe
  mutate(v1 = row.names(.)) %>%  # add row names as a column
  pivot_longer(cols = -v1, names_to = "v2", values_to = "cor") %>%        # reshape data
  filter(v1 != v2) %>%    # keep pairs that v1 matches v2, but are not the same
  drop_na() #remove the reverse correlations

#find the gene pairs mostly negatively correlated
neg_corr <- cor_paralogs_neuro_sim %>% 
  arrange(cor) %>% 
  head(n=10)

#visualise gene expression of negatively correlated pairs
  cpm_all_long %>% 
     mutate(pair = case_when(
    ensembl_gene_id  %in% neg_corr[1,] ~ "1",
    ensembl_gene_id  %in% neg_corr[2,] ~ "2",
    ensembl_gene_id  %in% neg_corr[3,] ~ "3",
    ensembl_gene_id  %in% neg_corr[4,] ~ "4",
    TRUE ~ NA)) %>% 
    drop_na %>%
    ggplot(aes(x=AGE, y= log_cpm, colour=gene_id)) +
  geom_point()+ 
    facet_wrap(~pair, scales = "free_y") +
  geom_smooth(se = FALSE) +
        scale_colour_manual(values=col) 
  
  
```


Questions:
Higher variance in <12?
Find pairs of switches (channel subunits etc). Plot their expression with age.
Enrichment analysis with ClusterProfiler (remember to combine into fewer pathways)
Enrichment for myelin-related genes in CC?
DE genes overlap with BD/SCZ GWAS?


