
title: "Import and normalise Maryland CC"
author: "Asbjørn"
date: "2022/2/24"
output: html_notebook
editor_options: 
  chunk_output_type: inline
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
library(tidyverse)
#library(annotables)
library(edgeR)
library(RColorBrewer)
library(ggrepel)
library(limma)
library(magrittr)
library(biomaRt)
```


This code is to import the Maryland data and produce normalised and filtered data.



### Add sample info

```{r add info}
info_OFC <- 
  readxl::read_xlsx("/Users/asbjorh/PhD/ANK3_HTS-amplicons/Maryland/data/Maryland_Samples_NBB1582DonorInfo_cleaned.xlsx", sheet = 1, col_names = TRUE) %>% 
  dplyr::select(c("RNA ID", "AGEYEARS", "AGEDAYS", "RIN AMG", "RNAseq", "PMINTERVAL", "CDEATHOFF")) %>% filter(RNAseq!=0) %>% select(-RNAseq) %>% distinct()

#simplify the names of samples to match data
info_OFC$`RNA ID` <-  str_remove(info_OFC$`RNA ID`, "MD_")
info_OFC$`RNA ID` <-  str_remove(info_OFC$`RNA ID`, "_RNA")
info_OFC$`RNA ID` <-  str_replace(info_OFC$`RNA ID`, "_", "-")
info_OFC <- dplyr::rename(info_OFC, sampleID="RNA ID")

info_OFC <- info_OFC %>% filter(!(sampleID=="5162-BA11" & `RIN AMG`== 9.6)) #this sample has 2 entries, unknown which was used for RNAseq, remove best RIN, keep RIN = 9.1

info_OFC <- dplyr::rename(info_OFC, RIN_AMG="RIN AMG")

info_OFC$sampleID <- str_remove(info_OFC$sampleID,  "-BA11")

info_CC <-
readxl::read_xlsx("~/proj_Maryland_WGCNA/data/RNAseq_CC samples_december_2021/sample_table_96_RNAseq_cleaned.xlsx", sheet = 1, col_names = TRUE)


info_CC <- 
  info_CC %>% 
  select(sampleID, RIN_AMG)

info_CC$sampleID <- as.character(info_CC$sampleID)

info_CC <- 
  info_CC %>% left_join(select(info_OFC, -RIN_AMG), by= "sampleID")


#change age (years+days) into decimal years. Remove cause of death.

info_CC <- 
  info_CC %>% 
  mutate(sampleID = paste0(sampleID, "-CC")) %>% 
  dplyr::mutate(AGE = AGEYEARS + (AGEDAYS/365)) %>%
    select(-AGEYEARS, -AGEDAYS)

info_OFC <- 
  info_OFC %>% 
  mutate(sampleID = paste0(sampleID, "-OFC")) %>% 
  select(-CDEATHOFF) %>% 
  dplyr::mutate(AGE = AGEYEARS + (AGEDAYS/365)) %>%
    select(-AGEYEARS, -AGEDAYS)

#mke a category for age. IA suggested +-12 Years, based on Kang et al (2011)'s definition of adolescence 

info_CC <-
  info_CC %>% 
  mutate(below12 = case_when(
    AGE < 12  ~ 1,
    TRUE ~ 0)
  )

info_CC$below12 <- as.factor(info_CC$below12)

#do the same for OFC
info_OFC <-
  info_OFC %>% 
  mutate(below12 = case_when(
    AGE < 12  ~ 1,
    TRUE ~ 0)
  )

info_OFC$below12 <- as.factor(info_OFC$below12)
```



```{r import counts CC}
#f = list.files("/Users/asbjorh/PhD/RNAseq_temporal/data/CC/Hughes-RNA6-2021-12-08/20_mapDataRSEM/",  pattern= ".genes.results$", full.names  = TRUE) #results from RSEM

f = list.files("/Users/asbjorh/PhD/RNAseq_temporal/data/CC/Hughes-RNA6-2021-12-08/40_countData/",  pattern= ".counts$", full.names  = TRUE) #results from 


#if RSEM results:
#dat = lapply(f, 
#             function(i){          #get the data from each sample
#  x = read_tsv(i, col_names = TRUE, skip = 0)
#  x = x[, c(1,4,5)] 
#  # add a column to say which file they're from
#  x$file = i
#  # Return your data
#  x
#}
#)

#if featurecounts results:
dat = lapply(f, 
             function(i){          #get the data from each sample
  x = read_tsv(i, col_names = FALSE, skip = 2)
  x = x[, c(1,6,7)] 
  # add a column to say which file they're from
  x$file = i
  # Return your data
  x
}
)


d<-
  do.call("rbind.data.frame", dat) #make a dataframe out of the imported datasets (combine by )


d$file <- #remove unnescessary prefix text from sample names
  str_remove(d$file, "/Users/asbjorh/PhD/RNAseq_temporal/data/CC/Hughes-RNA6-2021-12-08/40_countData//") 

d$file  <- #replace unnescessary suffix text from sample names
  str_replace(d$file , ".counts", "-CC") 


d <- 
  pivot_wider(dplyr::select(d, -X6), names_from= file, values_from = X7 ) 

 count <- d %>% column_to_rownames(var="X1")



#if using RSEM data
#d$file <- #remove unnescessary prefix text from sample names
 # str_remove(d$file, "/Users/asbjorh/PhD/RNAseq_temporal/data/CC/Hughes-RNA6-2021-12-08/20_mapDataRSEM//") 

#d$file  <- #replace unnescessary suffix text from sample names
 # str_replace(d$file , "_rsemExpr.genes.results", "-CC") 

#count <- d %>% select(gene_id, file, expected_count) %>% pivot_wider(names_from = file, values_from = expected_count, id_cols = gene_id) 
#count <- count %>% column_to_rownames(var="gene_id")


```

```{r import counts OFC}
f = list.files("/Users/asbjorh/PhD/RNAseq_temporal/data/OFC/",  pattern= ".counts$", full.names  = TRUE) #results from 


#if featurecounts results:
dat = lapply(f, 
             function(i){          #get the data from each sample
  x = read_tsv(i, col_names = FALSE, skip = 2)
  x = x[, c(1,6,7)] 
  # add a column to say which file they're from
  x$file = i
  # Return your data
  x
}
)


d<-
  do.call("rbind.data.frame", dat) #make a dataframe out of the imported datasets (combine by )


d$file <- #remove unnescessary prefix text from sample names
  str_remove(d$file, "/Users/asbjorh/PhD/RNAseq_temporal/data/OFC//") 

d$file  <- #replace unnescessary suffix text from sample names
  str_replace(d$file , ".counts", "-OFC") 


d <- 
  pivot_wider(dplyr::select(d, -X6), names_from= file, values_from = X7 ) 

 count <- d %>% column_to_rownames(var="X1")

 colnames(count) <- str_remove(colnames(count),  "-BA11")

```




```{r Process data}
clinMeta <- info_CC
clinMeta <- info_OFC

count <- count

SampleID <- colnames(count)[colnames(count) %in% clinMeta$sampleID] #make any other adjustments to which samples to include/ exclude here

data <- count[,colnames(count) %in% (SampleID)] #count is already in the form of a dataframe with rownames
data <- round(data) # round data since RSEM output may be non-integer

colData <- clinMeta


colData <- colData %>%  column_to_rownames(var="sampleID") #to make a dataframe with rownames
colData <- colData[colnames(data),] #order the sample names the same as the expression data
all( rownames(colData) == colnames(data))

#colData[,c(2,4,5,6)] <- lapply(colData[,c(2,4,5,6)],as.factor) #adjust any columns with factor values
#colData[,c(3,7,8)] <- lapply(colData[,c(3,7,8)],as.numeric) #similarly, make sure numerical values are properly labeled.


library(biomaRt)
ensembl <- useMart("ensembl", dataset="hsapiens_gene_ensembl", host="https://useast.ensembl.org")
genemap <- getBM(attributes=c("entrezgene_id","ensembl_gene_id", "external_gene_name", "gene_biotype", "chromosome_name", "start_position", "end_position"), filters="ensembl_gene_id", 
	values=rownames(data), mart=ensembl)
genemap <- genemap[ duplicated(genemap$ensembl_gene_id) == F,]
rownames(genemap) <- genemap$ensembl_gene_id


## Remove lowly expressed genes

library(edgeR)
keep <- filterByExpr(data#, design=colData$below12 #we don't want to have design here, do we? Because if genes are not at all expressed in either category it will be filtered out?
)
data.filt <- data[keep,]


## Keep only protein-coding and lncRNA genes
genemap2 <- genemap[rownames(data.filt),]
genemap2 <- subset(genemap2, gene_biotype %in% c("protein_coding","lncRNA"))
data.filt <- data.filt[rownames(genemap2),]
```


```{r DE analyses childhood vs adolescence}
analysis <- "CC_DE_below12"
analysis <- "OFC_DE_below12"

res_dir <- paste0("~/proj_Maryland_WGCNA/analysis/", analysis)


y <- DGEList(counts = data.filt, genes = genemap2, samples=SampleID)
y <- calcNormFactors(y)

design <- model.matrix(~ below12  + RIN_AMG + PMINTERVAL, data = colData)

v <- voomWithQualityWeights(y, design, plot = TRUE)
fit <- lmFit(v, design)
fit <- eBayes(fit)
summary(decideTests(fit, p.value=0.1))

res <- topTable(fit, coef="below121", n=Inf, p.value=1)
write.table(res, paste0(res_dir, "/de_limma_OFC_below12.txt"), sep="\t", row.names=F)
```


```{r plot DE genes}
#in order to visualise expression, we need the samples as rows and gene expression + covariates as columns
## Plot significant genes
library(ggsci)
library(ggpubr)

col <- get_palette("aaas", 10)

df <- cpm(y, log=TRUE, prior.count=3) #convert counts to normalised cpm
df <- df[rownames(head(res,10)), ] #only keep the top 10 significant genes

df <- df %>% t() %>% data.frame()     #have genes as columns
colnames(df) <- head(res,10)$external_gene_name #use gene symbols in stead of ensembl codes
df <- merge(df,colData.filt,by=0) 

df <- df %>% rename("SampleID" = Row.names )


#IA's plotting
# glist <- lapply(2:11, function (x) {
#   ggplot() + 
#     geom_boxplot(data=df,aes(x = below12, y = df[,x]), color ="black",size=0.7, outlier.shape=NA, fill="white") +
#     geom_point(data=df,aes(x = below12, y = df[,x], color = below12, fill=below12), position=position_jitterdodge(), pch=21,size=3.5, color="black", alpha=0.85) + 
#     labs(title=colnames(df)[x],x="",y="") +
#     scale_colour_manual(values=col[c(2,9,8)]) +
#     scale_fill_manual(values=col[c(2,9,8)]) +
#     theme_classic() +
#     theme(axis.text.x=element_text(size=20,color="black", angle=0, hjust=0.5),
#           axis.text.y=element_text(size=20,color="black"),
#           axis.title=element_text(size=16),
#           plot.title=element_text(size=20,hjust=0.5),
#           legend.text=element_text(size=20),
#           legend.title=element_blank(),
#           plot.margin=unit(c(0,0,0,0),"cm"),
#           panel.background=element_rect(linetype="solid", color="black", size=1)) 
# })
# 
# 
# annotate_figure(ggarrange(plotlist=glist,ncol=5,nrow=2,common.legend=T,legend="none"),
#                 left=text_grob("Normalized expression",size=20,rot=90))
# 

#prepare for flexible plotting
df_long <- 
  df %>%  pivot_longer(cols = 2:11, names_to = "gene_id", values_to = "log_cpm") 

df_long$below12 <- str_replace(df_long$below12, "1", "<12")
df_long$below12 <- str_replace(df_long$below12, "0", ">12")

df_long$below12 <- factor(df_long$below12, levels =  c("<12", ">12"))

df_long <- df_long %>%  rename("Age_group" = below12)

plot_expr <- df_long %>% ggplot(aes(x=Age_group, y= log_cpm)) + 
    geom_boxplot( color ="black",size=0.7, outlier.shape=NA, fill="white") +
    geom_point(aes( colour= Age_group, fill = Age_group), position= "jitter") +
  facet_wrap(~gene_id, scales = "free_y") +
      scale_colour_manual(values=col[c(1,6)]) +
    scale_fill_manual(values=col[c(1,6)]) +
  labs(title= paste0("Normalised expression, 10 significant DE genes, ", analysis)) +
  theme_bw()

ggsave(plot_expr, file = paste0(res_dir, "/plot_expr_top10sign.pdf" ))  
  

```




##################### BRAIN-GVEX #####################

## Process data
library(magrittr)

path <- "/Users/ibrahiaa/rnaseq_brain_data_sets/PsychENCODE/asbjørn_si_score/DE_OFC_5sets_TSD/"

clinMeta <- read.csv(paste0(path," info_GVEX.txt"), sep="\t", header=T)
count <- read.csv(paste0(path," count_GVEX.txt"), sep=",", header=T, check.names=F)
colnames(count)[1] <- "gene_id"

id <- subset(clinMeta, Study=="SMRI" & PrimaryDiagnosis %in% c("BD","HC") & Ethnicity =="CAUC")$SampleID %>% unique
data <- count[,c("gene_id",id)]

data$gene_id <- gsub("\\..*","",data$gene_id)
data <- data[ duplicated(data$gene_id) == F,]
rownames(data) <- data$gene_id
data <- data[,-1]
#data <- round(data) # round data since RSEM output may be non-integer

colData <- clinMeta
rownames(colData) <- colData$SampleID
colData <- colData[colnames(data),]
all( rownames(colData) == colnames(data))
colData[,c(2,4,5,6)] <- lapply(colData[,c(2,4,5,6)],as.factor)
colData[,c(3,7,8)] <- lapply(colData[,c(3,7,8)],as.numeric)
colData$PrimaryDiagnosis <- relevel(colData$PrimaryDiagnosis, ref="HC")

colData.filt <- subset(colData, SampleID %in% id)

data <- data[,rownames(colData.filt)]
all( rownames(colData.filt) == colnames(data))


library(biomaRt)
ensembl <- useMart("ensembl", dataset="hsapiens_gene_ensembl", host="useast.ensembl.org")
genemap <- getBM(attributes=c("entrezgene_id","ensembl_gene_id", "external_gene_name", "gene_biotype", "chromosome_name", "start_position", "end_position"), filters="ensembl_gene_id", 
	values=rownames(data), mart=ensembl)
genemap <- genemap[ duplicated(genemap$ensembl_gene_id) == F,]
rownames(genemap) <- genemap$ensembl_gene_id


## Remove lowly expressed genes

library(edgeR)
keep <- filterByExpr(data, design=colData.filt$PrimaryDiagnosis)
data.filt <- data[keep,]


## Keep only protein-coding and lncRNA genes
genemap2 <- genemap[rownames(data.filt),]
genemap2 <- subset(genemap2, gene_biotype %in% c("protein_coding","lncRNA"))
data.filt <- data.filt[rownames(genemap2),]


## DE analyses

library(limma)

y <- DGEList(counts = data.filt, genes = genemap2, samples=colData.filt)
y <- calcNormFactors(y)

design <- model.matrix(~ PrimaryDiagnosis + ReportedGender + Age + RIN + PMI, data = colData.filt)

v <- voomWithQualityWeights(y, design, plot = TRUE)
fit <- lmFit(v, design)
fit <- eBayes(fit)
summary(decideTests(fit, p.value=0.1))

res <- topTable(fit, coef="PrimaryDiagnosisBD", n=Inf, p.value=1)
write.table(res, "/Users/ibrahiaa/rnaseq_brain_data_sets/PsychENCODE/asbjørn_si_score/de_limma_gvex_smri.txt", sep="\t", row.names=F)


## Plot significant genes
library(ggsci)
library(ggpubr)

col <- get_palette("aaas", 10)

df <- cpm(y, log=TRUE, prior.count=3) #%>% t %>% data.frame
df <- df[ , rownames(head(res,10)) ]
colnames(df) <- head(res,10)$external_gene_name
df <- merge(df,colData.filt,by=0) 

glist <- lapply(2:11, function (x) {
  ggplot() + 
    geom_boxplot(data=df,aes(x = PrimaryDiagnosis, y = df[,x]), color ="black",size=0.7, outlier.shape=NA, fill="white") +
    geom_point(data=df,aes(x = PrimaryDiagnosis, y = df[,x], color = PrimaryDiagnosis, fill=PrimaryDiagnosis), position=position_jitterdodge(), pch=21,size=3.5, color="black", alpha=0.85) + 
    labs(title=colnames(df)[x],x="",y="") +
    scale_colour_manual(values=col[c(2,9,8)]) +
    scale_fill_manual(values=col[c(2,9,8)]) +
    theme_classic() +
    theme(axis.text.x=element_text(size=20,color="black", angle=0, hjust=0.5),
          axis.text.y=element_text(size=20,color="black"),
          axis.title=element_text(size=16),
          plot.title=element_text(size=20,hjust=0.5),
          legend.text=element_text(size=20),
          legend.title=element_blank(),
          plot.margin=unit(c(0,0,0,0),"cm"),
          panel.background=element_rect(linetype="solid", color="black", size=1)) 
})

annotate_figure(ggarrange(plotlist=glist,ncol=5,nrow=2,common.legend=T,legend="none"),
                left=text_grob("Normalized expression",size=20,rot=90))
